diff --git a/dlls/winex11.drv/x11drv_main.c b/dlls/winex11.drv/x11drv_main.c
index 2889fd216ff..2cc8df9ec84 100644
--- a/dlls/winex11.drv/x11drv_main.c
+++ b/dlls/winex11.drv/x11drv_main.c
@@ -191,7 +191,10 @@ const char * const X11DRV_atom_names[NB_XATOMS - FIRST_XATOM] =
     "text/plain",
     "text/rtf",
     "text/richtext",
-    "text/uri-list"
+    "text/uri-list",
+#ifdef __ANDROID__
+    "_NET_WM_HWND",
+#endif
 };

 /***********************************************************************
