diff --git a/dlls/ntdll/unix/sync.c b/dlls/ntdll/unix/sync.c
index bb7197cf247..0d8ccb828d1 100644
--- a/dlls/ntdll/unix/sync.c
+++ b/dlls/ntdll/unix/sync.c
@@ -60,9 +60,7 @@
 #ifdef HAVE_KQUEUE
 # include <sys/event.h>
 #endif
-#ifdef HAVE_LINUX_NTSYNC_H
-# include <linux/ntsync.h>
-#endif
+# include "ntsync_tmp.h"

 #include "ntstatus.h"
 #include "windef.h"
@@ -763,6 +761,11 @@ void close_inproc_sync( HANDLE handle )
 {
     struct inproc_sync *cache;

+    if (do_esync())
+    {
+        esync_close ( handle );
+        return;
+    }
     if (inproc_device_fd < 0) return;
     if ((cache = get_cached_inproc_sync( handle )))
     {
@@ -777,6 +780,8 @@ static NTSTATUS inproc_release_semaphore( HANDLE handle, ULONG count, ULONG *pre
 {
     struct inproc_sync stack, *sync;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_release_semaphore( handle, count, prev_count );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;
     if ((ret = get_inproc_sync( handle, INPROC_SYNC_SEMAPHORE, SEMAPHORE_MODIFY_STATE, &stack, &sync ))) return ret;
@@ -789,6 +794,8 @@ static NTSTATUS inproc_query_semaphore( HANDLE handle, SEMAPHORE_BASIC_INFORMATI
 {
     struct inproc_sync stack, *sync;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_query_semaphore( handle, info, NULL );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;
     if ((ret = get_inproc_sync( handle, INPROC_SYNC_SEMAPHORE, SEMAPHORE_QUERY_STATE, &stack, &sync ))) return ret;
@@ -801,6 +808,8 @@ static NTSTATUS inproc_set_event( HANDLE handle, LONG *prev_state )
 {
     struct inproc_sync stack, *sync;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_set_event( handle );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;
     if ((ret = get_inproc_sync( handle, INPROC_SYNC_EVENT, EVENT_MODIFY_STATE, &stack, &sync ))) return ret;
@@ -813,6 +822,8 @@ static NTSTATUS inproc_reset_event( HANDLE handle, LONG *prev_state )
 {
     struct inproc_sync stack, *sync;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_reset_event( handle );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;
     if ((ret = get_inproc_sync( handle, INPROC_SYNC_EVENT, EVENT_MODIFY_STATE, &stack, &sync ))) return ret;
@@ -825,6 +836,8 @@ static NTSTATUS inproc_pulse_event( HANDLE handle, LONG *prev_state )
 {
     struct inproc_sync stack, *sync;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_pulse_event( handle );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;
     if ((ret = get_inproc_sync( handle, INPROC_SYNC_EVENT, EVENT_MODIFY_STATE, &stack, &sync ))) return ret;
@@ -837,6 +850,8 @@ static NTSTATUS inproc_query_event( HANDLE handle, EVENT_BASIC_INFORMATION *info
 {
     struct inproc_sync stack, *sync;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_query_event( handle, info, NULL );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;
     if ((ret = get_inproc_sync( handle, INPROC_SYNC_EVENT, EVENT_QUERY_STATE, &stack, &sync ))) return ret;
@@ -849,6 +864,8 @@ static NTSTATUS inproc_release_mutex( HANDLE handle, LONG *prev_count )
 {
     struct inproc_sync stack, *sync;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_release_mutex( handle, prev_count );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;
     if ((ret = get_inproc_sync( handle, INPROC_SYNC_MUTEX, 0, &stack, &sync ))) return ret;
@@ -861,6 +878,8 @@ static NTSTATUS inproc_query_mutex( HANDLE handle, MUTANT_BASIC_INFORMATION *inf
 {
     struct inproc_sync stack, *sync;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_query_mutex( handle, info, NULL );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;
     if ((ret = get_inproc_sync( handle, INPROC_SYNC_MUTEX, MUTANT_QUERY_STATE, &stack, &sync ))) return ret;
@@ -902,6 +921,8 @@ static NTSTATUS inproc_wait( DWORD count, const HANDLE *handles, WAIT_TYPE type,
     struct inproc_sync *syncs[64], stack[ARRAY_SIZE(syncs)];
     int objs[ARRAY_SIZE(syncs)], alert_fd = 0;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_wait_objects( count, handles, type != WaitAll, alertable, timeout );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;

@@ -930,6 +951,8 @@ static NTSTATUS inproc_signal_and_wait( HANDLE signal, HANDLE wait,
     struct inproc_sync stack_signal, stack_wait, *signal_sync = &stack_signal, *wait_sync = &stack_wait;
     int alert_fd = 0;
     NTSTATUS ret;
+
+	if (do_esync()) return esync_signal_and_wait( signal, wait, alertable, timeout );

     if (inproc_device_fd < 0) return STATUS_NOT_IMPLEMENTED;

@@ -2424,6 +2447,12 @@ NTSTATUS WINAPI NtDelayExecution( BOOLEAN alertable, const LARGE_INTEGER *timeou
     /* if alertable, we need to query the server */
     if (alertable)
     {
+        if (do_esync())
+        {
+            NTSTATUS ret = esync_wait_objects( 0, NULL, TRUE, TRUE, timeout );
+            if (ret != STATUS_NOT_IMPLEMENTED)
+                return ret;
+        }
         /* Since server_wait will result in an unconditional implicit yield,
            we never return STATUS_NO_YIELD_PERFORMED */
         if ((status = server_wait( NULL, 0, SELECT_INTERRUPTIBLE | SELECT_ALERTABLE, timeout )) == STATUS_TIMEOUT)
