diff --git a/dlls/winex11.drv/keyboard.c b/dlls/winex11.drv/keyboard.c
index 65de84678ec..d2323e10ad5 100644
--- a/dlls/winex11.drv/keyboard.c
+++ b/dlls/winex11.drv/keyboard.c
@@ -68,6 +68,7 @@ WINE_DECLARE_DEBUG_CHANNEL(key);
 static const unsigned int ControlMask = 1 << 2;

 static int min_keycode, max_keycode, keysyms_per_keycode;
+static KeySym *key_mapping;
 static WORD keyc2vkey[256], keyc2scan[256];

 static int NumLockMask, ScrollLockMask, AltGrMask; /* mask in the XKeyEvent state */
@@ -1113,6 +1114,11 @@ static const WORD xfree86_vendor_key_vkey[256] =
     0, 0, 0, 0, 0, 0, 0, 0                                      /* 1008FFF8 */
 };

+static inline KeySym keycode_to_keysym( Display *display, KeyCode keycode, int index )
+{
+    return key_mapping[(keycode - min_keycode) * keysyms_per_keycode + index];
+}
+
 /* Returns the Windows virtual key code associated with the X event <e> */
 /* kbd_section must be held */
 static WORD EVENT_event_to_vkey( XIC xic, XKeyEvent *e)
@@ -1791,7 +1797,7 @@ static void init_keycode_mappings( Display *display )
 	      /* we seem to need to search the layout-dependent scancodes */
 	      int maxlen=0,maxval=-1,ok;
 	      for (i=0; i<syms; i++) {
-		keysym = XkbKeycodeToKeysym( display, keyc, 0, i );
+		keysym = keycode_to_keysym( display, keyc, i );
 		if ((keysym<0x8000) && (keysym!=' '))
                 {
                     if (!XkbTranslateKeySym(display, &keysym, 0, &ckey[i], 1, NULL))
@@ -1941,7 +1947,7 @@ static void init_keycode_mappings( Display *display )
     for (scan = 0x60, keyc = min_keycode; keyc <= max_keycode; keyc++)
       if (keyc2vkey[keyc]&&!keyc2scan[keyc]) {
 	const char *ksname;
-	keysym = XkbKeycodeToKeysym( display, keyc, 0, 0 );
+	keysym = keycode_to_keysym( display, keyc, 0 );
 	ksname = XKeysymToString(keysym);
 	if (!ksname) ksname = "NoSymbol";

@@ -1966,7 +1972,8 @@ void init_keyboard_layouts( Display *display )

     pthread_mutex_lock( &kbd_mutex );
     XDisplayKeycodes( display, &min_keycode, &max_keycode );
-    XFree( XGetKeyboardMapping( display, min_keycode, max_keycode + 1 - min_keycode, &keysyms_per_keycode ) );
+    if (key_mapping) XFree( key_mapping );
+    key_mapping = XGetKeyboardMapping( display, min_keycode, max_keycode + 1 - min_keycode, &keysyms_per_keycode );

     mmp = XGetModifierMapping( display );
     kcp = mmp->modifiermap;
@@ -1976,12 +1983,12 @@ void init_keyboard_layouts( Display *display )
         {
             for (int k = 0; *kcp && k < keysyms_per_keycode; k += 1)
             {
-                if (XkbKeycodeToKeysym( display, *kcp, 0, k ) == XK_Num_Lock)
+                if (keycode_to_keysym( display, *kcp, k ) == XK_Num_Lock)
                 {
                     NumLockMask = 1 << i;
                     TRACE_(key)( "NumLockMask is %x\n", NumLockMask );
                 }
-                else if (XkbKeycodeToKeysym( display, *kcp, 0, k ) == XK_Scroll_Lock)
+                else if (keycode_to_keysym( display, *kcp, k ) == XK_Scroll_Lock)
                 {
                     ScrollLockMask = 1 << i;
                     TRACE_(key)( "ScrollLockMask is %x\n", ScrollLockMask );
@@ -2172,7 +2179,7 @@ SHORT X11DRV_VkKeyScanEx( WCHAR wChar, HKL hkl )
     }

     for (index = 0; index < 4; index++) /* find shift state */
-        if (XkbKeycodeToKeysym( display, keycode, 0, index ) == keysym) break;
+        if (keycode_to_keysym( display, keycode, index ) == keysym) break;

     pthread_mutex_unlock( &kbd_mutex );

@@ -2422,7 +2429,7 @@ INT X11DRV_GetKeyNameText( LONG lParam, LPWSTR lpBuffer, INT nSize )
       INT rc;

       keyc = (KeyCode) keyi;
-      keys = XkbKeycodeToKeysym( display, keyc, 0, 0 );
+      keys = keycode_to_keysym( display, keyc, 0 );
       name = XKeysymToString(keys);

       if (name && (vkey == VK_SHIFT || vkey == VK_CONTROL || vkey == VK_MENU))
