diff --git a/server/thread.c b/server/thread.c
index 7207f918400..babe3f05ad7 100644
--- a/server/thread.c
+++ b/server/thread.c
@@ -1773,7 +1773,12 @@ DECL_HANDLER(init_first_thread)
     set_reply_data( supported_machines,
                     min( supported_machines_count * sizeof(unsigned short), get_reply_max_size() ));
 
-    if ((fd = get_inproc_device_fd()) >= 0)
+    if (do_esync())
+    {
+        reply->inproc_device = ESYNC_USED_BY_SERVER;
+        send_client_fd(process, esync_get_shm_fd(), reply->inproc_device);
+    }
+    else if ((fd = get_inproc_device_fd()) >= 0)
     {
         reply->inproc_device = get_process_id( process ) | 1;
         send_client_fd( process, fd, reply->inproc_device );
